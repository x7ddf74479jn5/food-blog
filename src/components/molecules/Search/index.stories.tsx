import type { ComponentMeta, ComponentStory } from "@storybook/react";

import Search from ".";

export default {
  title: "Molecules/Search",
  component: Search,
} as ComponentMeta<typeof Search>;

const Template: ComponentStory<typeof Search> = (args) => <Search {...args} />;

export const Default = Template.bind({});
