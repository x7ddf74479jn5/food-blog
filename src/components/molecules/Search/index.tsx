import { useRouter } from "next/router";
import { useEffect, useRef, useState } from "react";
import { IconContext } from "react-icons";
import { FaSearch } from "react-icons/fa";

import { useSearchHistoryContext } from "@/context";
import { urlTable } from "@/utils/paths/url";

const Search: React.FC = () => {
  const ref = useRef<HTMLDivElement>(null);
  const [query, setQuery] = useState("");
  const [isActive, setIsActive] = useState(false);

  const router = useRouter();
  const { histories, setHistories } = useSearchHistoryContext();

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => setQuery(event.target.value);

  const handleFocus = () => setIsActive(true);

  const search = (q: string) => {
    router.push(
      {
        pathname: urlTable.search,
        query: { q },
      },
      undefined,
      { shallow: true }
    );

    if (!histories.includes(query)) {
      setHistories((prev) => [query, ...prev].slice(0.5));
    }

    setQuery("");
    setIsActive(false);
  };

  const handleKeyPress = (event: React.KeyboardEvent<HTMLInputElement>) => {
    if (query === "") {
      return;
    }

    if (event.key === "Enter") {
      search(query);
    }
  };

  const handleSubmit = () => {
    if (query === "") {
      return;
    }

    search(query);
  };

  const handleClickInside = (history: string) => {
    search(history);
  };

  useEffect(() => {
    const handleClickOutside = (event: Event) => {
      if (ref.current && !ref.current.contains(event.target as Node)) {
        setIsActive(false);
      }
    };
    document.addEventListener("mousedown", handleClickOutside);
    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  });

  return (
    <div ref={ref} className="relative">
      <div className="flex items-center py-2 px-3 space-x-2 bg-gray-100 dark:bg-gray-700 rounded-full">
        <IconContext.Provider
          value={{
            color: "gray",
          }}
        >
          <FaSearch />
        </IconContext.Provider>

        <input
          type="search"
          placeholder="Search..."
          className="w-full dark:text-gray-100 bg-gray-100 dark:bg-gray-700 focus:outline-none"
          onChange={handleChange}
          onFocus={handleFocus}
          onKeyPress={handleKeyPress}
          onSubmit={handleSubmit}
          value={query}
          inputMode="text"
        />
      </div>
      {isActive && histories && histories.length > 0 && (
        <ul className="overflow-y-auto fixed z-10 py-2 mt-1 mr-4 max-h-[50vh] bg-white dark:bg-gray-700 rounded-md border border-gray-200 dark:border-gray-600 shadow-lg">
          {histories.map((history) => (
            <li
              key={history}
              className="py-2 px-4 text-sm leading-5 text-left dark:text-gray-100 hover:bg-gray-100 dark:hover:bg-green-500 cursor-pointer"
            >
              <button onClick={() => handleClickInside(history)}>{history}</button>
            </li>
          ))}
        </ul>
      )}
    </div>
  );
};

export default Search;
