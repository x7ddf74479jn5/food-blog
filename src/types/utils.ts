export type TCategoryOrTagColor = "green" | "blue" | "indigo" | "purple" | "red" | "yellow" | "pink" | "gray";

export type TImage = {
  url: string;
  width: number;
  height: number;
  blurDataURL?: string;
};

export type TImageOption = {
  fontColor: string;
};

export type ValueOf<T> = T[keyof T];

export type MappedConst<T extends string> = {
  [K in T]: K;
};
