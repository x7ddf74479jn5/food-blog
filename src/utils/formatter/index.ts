export { formatPageTitle, formatPageUrl } from "./formatPage";
export { getBGColor } from "./getColor";
export { getExcerpt } from "./getExcerpt";
