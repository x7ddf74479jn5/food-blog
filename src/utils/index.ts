export const toIdleTask = (callback: () => void) => {
  if ("requestIdleCallback" in window) {
    requestIdleCallback(callback);
  } else {
    setTimeout(callback, 0);
  }
};
