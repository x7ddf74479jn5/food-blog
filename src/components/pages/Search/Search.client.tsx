"use client";

import type { MicroCMSQueries } from "microcms-js-sdk";
import { useSearchParams } from "next/navigation";
import { useEffect, useState } from "react";

import { DefaultLayout } from "@/components/layouts";
import { ArticleSWRContainer } from "@/components/model/article";
import type { TConfig } from "@/types";
import { formatPageTitle, formatPageUrl } from "@/utils/formatter";
import { getBackLinks, urlTable } from "@/utils/paths/url";

const useSearchPage = (config: TConfig) => {
  const [heading, setHeading] = useState("");
  const params = useSearchParams();
  const q = params.get("q") ?? "";
  const { host, siteTitle } = config;
  const pageTitle = formatPageTitle(heading, siteTitle);
  const url = formatPageUrl(`${urlTable.search}/q=${q}`, host);

  useEffect(() => {
    setHeading(`検索結果：${q}`);
  }, [q]);

  return {
    heading,
    pageTitle,
    url,
  };
};

export const useQueryOption = () => {
  const queryFilters = [];
  const params = useSearchParams();
  const q = params.get("q") ?? "";
  const categoryParam = params.get("category");
  const tagsParam = params.get("tags");

  if (categoryParam) {
    const searchCategory = `categories[equals]${categoryParam}`;
    queryFilters.push(searchCategory);
  }

  if (tagsParam) {
    const searchTags = tagsParam
      .split(",")
      .map((tag) => `tags[contains]${tag}`)
      .join("[and]");
    queryFilters.push(searchTags);
  }

  const filters = queryFilters.length > 1 ? queryFilters.join("[and]") : queryFilters.join("");
  const queryOptions: MicroCMSQueries = { filters, q };

  return queryOptions;
};

type Props = {
  config: TConfig;
};
export const SearchClient: React.FC<Props> = ({ config }) => {
  const { heading, pageTitle, url } = useSearchPage(config);
  const queryOptions = useQueryOption();
  const backLinks = getBackLinks([urlTable.home, urlTable.categories]);

  return (
    <DefaultLayout pageTitle={pageTitle} url={url} backLinks={backLinks}>
      <h1 className="mb-8">{heading}</h1>
      <div className="w-full">
        <ArticleSWRContainer queryOptions={queryOptions} />
      </div>
    </DefaultLayout>
  );
};
