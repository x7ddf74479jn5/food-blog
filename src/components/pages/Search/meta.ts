import { cache } from "react";

import { fetchConfig } from "@/repositories";
import type { PageMeta } from "@/types";
import { formatPageTitle, formatPageUrl } from "@/utils/formatter";
import { urlTable } from "@/utils/paths/url";

export const getSearchPageMeta = cache(async (): Promise<PageMeta> => {
  const { host, siteDescription, siteImage, siteTitle } = await fetchConfig();
  const title = formatPageTitle("検索", siteTitle);
  const url = formatPageUrl(`${urlTable.search}`, host);

  return { description: siteDescription, image: siteImage.url, title, url };
});
