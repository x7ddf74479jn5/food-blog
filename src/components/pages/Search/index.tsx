import type { MicroCMSQueries } from "microcms-js-sdk";
import { useRouter } from "next/router";
import type { ParsedUrlQuery } from "querystring";
import { useEffect, useState } from "react";

import { HtmlHeadBase } from "@/components/functions/meta";
import { DefaultLayout } from "@/components/layouts";
import { ArticleSWRContainer } from "@/components/organisms/article";
import type { TCategory, TConfig, TPickup, TRankedArticle, TTag } from "@/types";
import { formatPageTitle, formatPageUrl } from "@/utils/formatter";
import { getBackLinks, urlTable } from "@/utils/paths/url";

const useSearchPage = (config: TConfig, query: ParsedUrlQuery) => {
  const [heading, setHeading] = useState("");
  const router = useRouter();
  const q = query.q ? String(query.q) : "";
  const { host, siteTitle } = config;
  const pageTitle = formatPageTitle(heading, siteTitle);
  const url = formatPageUrl(`${urlTable.search}/q=${q ?? ""}`, host);

  useEffect(() => {
    if (router.isReady) {
      setHeading(`検索結果：${q ?? ""}`);
    }
  }, [q, router.isReady]);

  return {
    heading,
    pageTitle,
    url,
  };
};

export const useQueryOption = (query: ParsedUrlQuery) => {
  const queryFilters = [];

  const q = query.q ? String(query.q) : "";

  if (query.category) {
    const searchCategory = `categories[equals]${String(query.category)}`;
    queryFilters.push(searchCategory);
  }

  if (query.tags) {
    const searchTags = String(query.tags)
      .split(",")
      .map((tag) => `tags[contains]${tag}`)
      .join("[and]");
    queryFilters.push(searchTags);
  }

  const filters = queryFilters.length > 1 ? queryFilters.join("[and]") : queryFilters.join("");
  const queryOptions: MicroCMSQueries = { filters, q };

  return queryOptions;
};

export type SearchProps = {
  config: TConfig;
  categories: TCategory[];
  tags: TTag[];

  pickup: TPickup;
  popularArticles: TRankedArticle[];
};

export const Search: React.FC<SearchProps> = ({ categories, config, pickup, popularArticles, tags }) => {
  const { host } = config;
  const { query } = useRouter();
  const { heading, pageTitle, url } = useSearchPage(config, query);
  const queryOptions = useQueryOption(query);
  const backLinks = getBackLinks([urlTable.home, urlTable.categories]);

  return (
    <DefaultLayout
      config={config}
      pageTitle={pageTitle}
      url={url}
      backLinks={backLinks}
      categories={categories}
      tags={tags}
      pickup={pickup}
      popularArticles={popularArticles}
    >
      <HtmlHeadBase indexUrl={host} pageTitle={pageTitle} url={url} />
      <div className="mb-8">
        <h1>{heading}</h1>
      </div>
      <div className="w-full">
        <ArticleSWRContainer queryOptions={queryOptions} />
      </div>
    </DefaultLayout>
  );
};
