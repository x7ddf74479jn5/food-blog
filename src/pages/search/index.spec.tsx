import { mockCategories, mockConfig, mockPickup } from "@mocks/data";
import type { NextRouter } from "jest/test-utils";
import { render, screen, withMockedRouter } from "jest/test-utils";
import { server } from "mocks/msw/server";
import renderer from "react-test-renderer";

import { formatPageTitle } from "@/utils/formatter";

import Search, { getStaticProps } from "./index.page";

beforeAll(() => server.listen());
afterAll(() => server.close());

jest.mock("next/head", () => {
  return {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    __esModule: true,
    default: ({ children }: { children: Array<React.ReactElement> }) => {
      return <>{children}</>;
    },
  };
});

describe("pages/search", () => {
  const mockCategoryList = Object.values(mockCategories);
  const mockCategoryRice = mockCategories.rice;
  const mockRouter: Partial<NextRouter> = {
    query: {
      q: "keyword",
    },
  };
  it("snapshot", () => {
    const tree = renderer
      .create(
        withMockedRouter(mockRouter, <Search categories={mockCategoryList} config={mockConfig} pickup={mockPickup} />)
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  // FIXME: sdkとnextをアップグレードしたら壊れた
  it.skip("getStaticProps", async () => {
    const result = await getStaticProps({
      params: {
        slug: mockCategoryRice.slug,
      },
    });

    if ("props" in result) {
      const { categories, config, pickup } = result.props;
      expect(categories).toStrictEqual(mockCategoryList);
      expect(pickup).toStrictEqual(mockPickup);
      expect(config).toStrictEqual(mockConfig);
    }
  });

  it("OK: 初期レンダリング", async () => {
    const { unmount } = render(
      withMockedRouter(mockRouter, <Search categories={mockCategoryList} config={mockConfig} pickup={mockPickup} />)
    );
    const expectedHeading = `検索結果：keyword`;
    const h1 = screen.getByRole("heading", { level: 1 });
    expect(h1).toHaveTextContent(expectedHeading);
    const expectedTitle = formatPageTitle(expectedHeading, mockConfig.siteTitle);
    expect(document.title).toBe(expectedTitle);
    unmount();
  });
});
