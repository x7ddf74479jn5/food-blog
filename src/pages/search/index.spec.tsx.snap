// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/search snapshot 1`] = `
<div
  className="flex flex-col px-4 mx-auto mb-16 max-w-screen-lg min-h-screen"
>
  <header
    className="py-2"
  >
    <div
      className="flex justify-between items-center"
    >
      <a
        className="text-2xl font-bold text-green-600"
        href="/"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Food Blog
      </a>
      <div
        className="hidden sm:block w-1/3"
      >
        <div
          className="relative"
        >
          <div
            className="flex items-center py-2 px-3 space-x-2 bg-gray-100 dark:bg-gray-700 rounded-full"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": "gray",
                }
              }
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
              />
            </svg>
            <input
              className="w-full dark:text-gray-100 bg-gray-100 dark:bg-gray-700 focus:outline-none"
              inputMode="text"
              onChange={[Function]}
              onFocus={[Function]}
              onKeyPress={[Function]}
              onSubmit={[Function]}
              placeholder="Search..."
              type="search"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        className="flex items-center space-x-4"
      >
        <div
          className="inline-block relative text-left"
        >
          <span
            className="rounded-md shadow-sm"
          >
            <button
              aria-expanded={false}
              aria-haspopup={true}
              className="inline-flex justify-center py-1 px-3 w-full text-sm font-medium leading-5 text-gray-700 hover:text-gray-500 dark:text-gray-100 bg-white dark:bg-gray-700 rounded-md border border-gray-300 dark:border-gray-500 dark:hover:border-gray-400 transition duration-150 ease-in-out focus:outline-none"
              id="headlessui-menu-button-:r0:"
              onClick={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              type="button"
            >
              <span>
                カテゴリー
              </span>
              <svg
                className="-mr-1 ml-2 w-5 h-5"
                fill="currentColor"
                viewBox="0 0 20 20"
              >
                <path
                  clipRule="evenodd"
                  d="M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z"
                  fillRule="evenodd"
                />
              </svg>
            </button>
          </span>
        </div>
      </div>
    </div>
    <div
      className="block sm:hidden mt-2"
    >
      <div
        className="relative"
      >
        <div
          className="flex items-center py-2 px-3 space-x-2 bg-gray-100 dark:bg-gray-700 rounded-full"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": "gray",
              }
            }
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            />
          </svg>
          <input
            className="w-full dark:text-gray-100 bg-gray-100 dark:bg-gray-700 focus:outline-none"
            inputMode="text"
            onChange={[Function]}
            onFocus={[Function]}
            onKeyPress={[Function]}
            onSubmit={[Function]}
            placeholder="Search..."
            type="search"
            value=""
          />
        </div>
      </div>
    </div>
  </header>
  <div
    className="flex flex-col gap-12 justify-center items-center mt-8"
    role="alert"
  >
    <svg
      className="w-32 h-32 text-gray-500"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 496 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 448c-110.3 0-200-89.7-200-200S137.7 56 248 56s200 89.7 200 200-89.7 200-200 200zm129.1-303.8c-3.8-4.4-10.3-5.4-15.3-2.5l-80 48c-3.6 2.2-5.8 6.1-5.8 10.3s2.2 8.1 5.8 10.3l80 48c5.4 3.2 11.8 1.6 15.3-2.5 3.8-4.5 3.9-11 .1-15.5L343.6 208l33.6-40.3c3.8-4.5 3.7-11.1-.1-15.5zM220 208c0-4.2-2.2-8.1-5.8-10.3l-80-48c-5-3-11.5-1.9-15.3 2.5-3.8 4.5-3.9 11-.1 15.5l33.6 40.3-33.6 40.3c-3.8 4.5-3.7 11 .1 15.5 3.5 4.1 9.9 5.7 15.3 2.5l80-48c3.6-2.2 5.8-6.1 5.8-10.3zm28 64c-45.4 0-100.9 38.3-107.8 93.3-1.5 11.8 6.9 21.6 15.5 17.9C178.4 373.5 212 368 248 368s69.6 5.5 92.3 15.2c8.5 3.7 17-6 15.5-17.9-6.9-55-62.4-93.3-107.8-93.3z"
      />
    </svg>
    <div
      className="flex flex-col gap-8 mx-auto text-center"
    >
      <h1
        className="flex justify-center text-4xl font-bold"
      >
        Something went wrong
      </h1>
      <p>
        サイト上で問題が発生しました。
      </p>
    </div>
    <button
      className="flex flex-row gap-2 justify-center items-center py-2 px-4 text-white bg-green-700 hover:bg-opacity-90 rounded-xl border-2 border-green-600"
      onClick={[Function]}
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 320 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
        />
      </svg>
      <span>
        Homeに戻る
      </span>
    </button>
  </div>
</div>
`;
